Mathlib/Analysis/LocallyConvex/AbsConvex.lean

```diff
@@ -70,8 +161,201 @@ theorem nhds_basis_abs_convex_open :
+      · ring_nf
+      · rw [smul_neg, ← sub_eq_add_neg, ← sub_smul]
+        apply congrFun (congrArg HSMul.hSMul _) y
+        ring_nf
+
+@[simp]
+theorem convexHull_union_neg_eq_absConvexHull {s : Set E} :
+    convexHull ℝ (s ∪ -s) = absConvexHull ℝ s := by
+  rw [absConvexHull_eq_convexHull_balancedHull]
+  exact le_antisymm (convexHull_mono (union_subset (subset_balancedHull ℝ)
+    (fun _ _ => by rw [mem_balancedHull_iff]; use -1; aesop)))
+    (by
+      rw [← Convex.convexHull_eq (convex_convexHull ℝ (s ∪ -s))]
+      exact convexHull_mono balancedHull_subset_convexHull_union_neg)
+
+
+lemma add_self_eq_smul_two {V : Set E} (h : Convex ℝ V) : V + V = (2 : ℝ) • V := by
+  rw [← one_add_one_eq_two, Convex.add_smul h (zero_le_one' ℝ) (zero_le_one' ℝ), MulAction.one_smul]
+
+variable (E 𝕜) {s : Set E}
+variable [NontriviallyNormedField 𝕜]  [Module 𝕜 E]
+variable [SMulCommClass ℝ 𝕜 E]
+variable [TopologicalSpace E] [TopologicalAddGroup E]  [lcs : LocallyConvexSpace ℝ E]
+  [ContinuousSMul ℝ E]
+
+-- TVS II.25 Prop3
+theorem totallyBounded_absConvexHull
+    (hs : TotallyBounded (uniformSpace := TopologicalAddGroup.toUniformSpace E) s) :
+    TotallyBounded (uniformSpace := TopologicalAddGroup.toUniformSpace E) (absConvexHull ℝ s) := by
+  intro d' hd'
+  letI := TopologicalAddGroup.toUniformSpace E
+  obtain ⟨d,⟨⟨N,⟨hN₁,hN₂⟩⟩, hd₂⟩⟩ := comp_mem_uniformity_sets hd'
+  obtain ⟨V,⟨hS₁,hS₂,hS₃⟩⟩ := (locallyConvexSpace_iff_exists_absconvex_subset_zero E).mp lcs N hN₁
+  let d₂ := {(x,y) | y-x ∈ V}
+  obtain ⟨t,⟨htf,hts⟩⟩ := hs d₂ (by
+    rw [uniformity_eq_comap_nhds_zero' E] -- d₂ is a uniformity
+    aesop)
+  have s1 : SymmetricRel d₂ := by
+    ext ⟨x,y⟩
+    simp only [mem_preimage, Prod.swap_prod_mk]
+    constructor
+    · intro h
+      simp only [mem_setOf_eq, d₂] at h
+      rw [← Balanced.neg_mem_iff hS₂.1, neg_sub] at h
+      exact h
+    · intro h
+      simp only [mem_setOf_eq, d₂] at h
+      rw [← Balanced.neg_mem_iff hS₂.1, neg_sub] at h
+      exact h
+  have e1 (y : E) : {x | (x, y) ∈ d₂} = y +ᵥ V := by
+    rw [← UniformSpace.ball_eq_of_symmetry s1, ← uniform_space_ball_eq_vadd]
+    rfl
+  have e2 {t₁ : Set E} : ⋃ y ∈ t₁, {x | (x, y) ∈ d₂} = t₁ + V := by
+    aesop
+  rw [e2] at hts
+  have e4 : (absConvexHull ℝ) s ⊆ (convexHull ℝ) (t ∪ -t) + V := by
+    rw [convexHull_union_neg_eq_absConvexHull (s := t), ← AbsConvex.absConvexHull_eq hS₂]
+    exact le_trans (absConvexHull_mono hts) (absConvexHull_add_subset ℝ)
+  have e6 : TotallyBounded (uniformSpace := TopologicalAddGroup.toUniformSpace E)
+      ((convexHull ℝ) (t ∪ -t)) := IsCompact.totallyBounded
+        (Set.Finite.isCompact_convexHull (finite_union.mpr ⟨htf,Finite.neg htf⟩))
+  obtain ⟨t',⟨htf',hts'⟩⟩ := e6 d₂ (by
+    rw [uniformity_eq_comap_nhds_zero']
+    aesop
+  )
+  rw [e2] at hts'
+  have e7: (absConvexHull ℝ) s ⊆ t' + (2 : ℝ) • V := by
+    rw [← add_self_eq_smul_two hS₂.2, ← add_assoc]
+    exact le_trans e4 (Set.add_subset_add_right hts')
```

Comment:
I think you'd make your life easier by just using `V + V` directly. Of course both are equal, but introducing smul here can only cause harm I think.

Reply:
Thanks - that's much better - in fact gives us a more general result for `V+W`.